contracts/cw20-streams/src/contract.rs

```diff
@@ -0,0 +1,446 @@
+    }
+
+    let validated_recipient = deps.api.addr_validate(recipient.as_str())?;
+    if validated_recipient != recipient {
+        return Err(ContractError::InvalidRecipient {});
+    }
+
+    let config = CONFIG.load(deps.storage)?;
+    if config.owner == recipient {
+        return Err(ContractError::InvalidRecipient {});
+    }
+
+    if start_time > end_time {
+        return Err(ContractError::InvalidStartTime {});
+    }
+
+    let block_time = env.block.time.seconds();
+    if start_time < block_time {
+        return Err(ContractError::InvalidStartTime {});
+    }
+
+    let duration: Uint128 = end_time.checked_sub(start_time).unwrap().into();
+
+    if amount < duration {
+        return Err(ContractError::InvalidDuration {});
+    }
+
+    if amount.u128().checked_rem(duration.u128()).unwrap() != 0 {
+        return Err(ContractError::InvalidDuration {});
+    }
+
+    let rate_per_second: Uint128 = amount.u128().checked_div(duration.u128()).unwrap().into();
+
+    let stream = Stream {
+        owner: validated_owner,
+        recipient: validated_recipient,
+        amount,
+        claimed_amount: Uint128::zero(),
+        start_time,
+        end_time,
+        rate_per_second,
+    };
+    save_stream(deps, &stream)?;
+
+    Ok(Response::new()
+        .add_attribute("method", "try_create_stream")
+        .add_attribute("owner", owner)
+        .add_attribute("recipient", recipient)
+        .add_attribute("amount", amount)
+        .add_attribute("start_time", start_time.to_string())
+        .add_attribute("end_time", end_time.to_string()))
+}
+
+pub fn execute_receive(
+    env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    wrapped: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if config.cw20_addr != info.sender {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let msg: ReceiveMsg = from_binary(&wrapped.msg)?;
+    match msg {
+        ReceiveMsg::CreateStream {
+            recipient,
+            start_time,
+            end_time,
+        } => try_create_stream(
+            env,
+            deps,
+            wrapped.sender,
+            recipient,
+            wrapped.amount,
+            start_time,
+            end_time,
+        ),
+    }
+}
+
+pub fn try_withdraw(
+    env: Env,
+    deps: DepsMut,
+    info: MessageInfo,
+    id: u64,
+) -> Result<Response, ContractError> {
+    let mut stream = STREAMS.load(deps.storage, id)?;
+    if stream.recipient != info.sender {
+        return Err(ContractError::NotStreamRecipient {});
+    }
+
+    if stream.claimed_amount >= stream.amount {
+        return Err(ContractError::StreamFullyClaimed {});
+    }
+
+    let block_time = env.block.time.seconds();
+    if stream.start_time >= block_time {
+        return Err(ContractError::StreamNotStarted {});
+    }
+
+    let unclaimed_amount = u128::from(block_time)
+        .checked_sub(stream.start_time.into())
+        .unwrap()
```

Comment:
don't use `unwrap()` in smart contract code. use `?` instead

Reply:
I agree strongly with removing `unwrap()`.

However, I think this will need 2 `Into::into()` calls to work, so maybe `?` will not work. If you change this to `?` and still get an error (probably why there is `unwrap()` here, try:

`.map_err(|e| ContractError::Std(e.into())?` a little longer but should work.

And you can extract 
`fn math_error(e: OverflowError) -> ContractError { ContractError::Std(e.into()) }` 
if you want to make these lines simpler:
`.map_err(math_error)?`

Reply:
Great suggestion. I ended up going with the saturating_sub idea @ethanfrey suggested, so these checked_sub calls were no longer needed. Removed the uses of unwrap in other parts of the code as well